### ISSUE TITLE
IndexError: Out of bounds on buffer access (axis 0) -- Using FastscapeEroder component

### ISSUE THREAD

**Issue by amanaster2** [landlab/landlab#507]
Hi!

As I was trying to use the FastscapeEroder component earlier, the following error popped up:
![image](https://user-images.githubusercontent.com/26778806/27810489-da420274-600f-11e7-8ede-e787ab6dd7da.png)

I tried uninstalling and reinstalling the developer version of Landlab (per @saisiddu 's suggestion) to no avail. Is there a known issue with this .pyx file? Could my code be causing the error? Any suggestions are appreciated!!

Thank you,
Amanda

**Comment by jadams15** [comment]
Hi Amanda! Can you try to recompile? From the terminal, change directories to the top level "landlab" folder and try to run the following line of code:

`python setup.py develop`

and see if that helps!

**Comment by amanaster2** [comment]
Hi Jordan! I've already recompiled, and unfortunately I haven't had any luck.

**Comment by kbarnhart** [comment]
@amanaster2 I most recently worked with the .pyx file you are having issues with. Could you create a minimally complete example that reproduces the error and post it (tagging me)? I'll then look into what is causing the error.

**Comment by amanaster2** [comment]
@kbarnhart : Below is the basic idea of the script I'm using. It's very possible that I'm incorrectly using FastscapeEroder as this is the first time I'm implementing it!

```
from landlab.components import KinwaveImplicitOverlandFlow, FastscapeEroder
from landlab.io import read_esri_ascii
import numpy as np

#load DEM for grid -- this is just the DEM I used
(mg, z) = read_esri_ascii(r'C:\Users\Amanda\Desktop\Lidar_amanda\Lidar_amanda\lidar_new.txt', name='topographic__elevation')

#set boundary conditions
mg.set_nodata_nodes_to_closed(z, -9999)
outlet_id = mg.core_nodes[np.argmin(mg.at_node['topographic__elevation'][mg.core_nodes])]                     
mg.set_watershed_boundary_condition_outlet_id(outlet_id, z) 

#initialize components
knwv = KinwaveImplicitOverlandFlow(mg, runoff_rate = 0.0, roughness = 0.03, depth_exp = 1.6666667)
sp = FastscapeEroder(mg, K_sp = 1.0)

#set initial conditions/constants
elapsed_time = 0
model_run_time = 7200 #seconds
storm_duration = 3600 #seconds
dt = 10. #seconds
rr = 10*2.77778E-07 #runoff rate in m/s
hydrograph_time = [0]
discharge_at_outlet = [0]
    
#run the model
while elapsed_time <= model_run_time:
    if elapsed_time < storm_duration:
        knwv.run_one_step(dt, current_time = elapsed_time, runoff_rate = rr)
        sp.run_one_step(dt = dt, rainfall_intensity_if_used = rr)
    else:
        knwv.run_one_step(dt, current_time = elapsed_time, runoff_rate = 0.0)
        sp.run_one_step(dt = dt, rainfall_intensity_if_used = 0.0)
    
    #what is the discharge at the outlet?    
    q_at_outlet = mg.at_node['surface_water_inflow__discharge'][outlet_id]

    hydrograph_time.append(elapsed_time/3600.)
    discharge_at_outlet.append(q_at_outlet)
                          
    elapsed_time += dt
```

**Comment by kbarnhart** [comment]
@amanaster2 

Could you try and reproduce the error with a script that uses a synthetic initial topography and then post that script?

**Comment by amanaster2** [comment]
@kbarnhart : The following code still reproduces the error.

```
from landlab.components import KinwaveImplicitOverlandFlow, FastscapeEroder
from landlab import RasterModelGrid
import numpy as np

#create 100x100 grid and add a slanted surface elevation
init = np.ones([10000])
surface = init*10
mg = RasterModelGrid(100, 100, spacing = (1,1))
z = mg.add_field('topographic__elevation', surface + mg.node_y*0.05, at = 'node')

#set boundary conditions
mg.set_nodata_nodes_to_closed(z, -9999)
outlet_id = mg.core_nodes[np.argmin(mg.at_node['topographic__elevation'][mg.core_nodes])]                     
mg.set_watershed_boundary_condition_outlet_id(outlet_id, z) 

#initialize components
knwv = KinwaveImplicitOverlandFlow(mg, runoff_rate = 0.0, roughness = 0.03, depth_exp = 1.6666667)
sp = FastscapeEroder(mg, K_sp = 1.0)

#set initial conditions/constants
elapsed_time = 0
model_run_time = 7200 #seconds
storm_duration = 3600 #seconds
dt = 10. #seconds
hydrograph_time = [0]
discharge_at_outlet = [0]
rr = 10*2.77778E-07 #runoff rate in m/s
    
#run the model
while elapsed_time <= model_run_time:
    if elapsed_time < storm_duration:
        knwv.run_one_step(dt, current_time = elapsed_time, runoff_rate = rr)
        sp.run_one_step(dt = dt, rainfall_intensity_if_used = rr)
    else:
        knwv.run_one_step(dt, current_time = elapsed_time, runoff_rate = 0.0)
        sp.run_one_step(dt = dt, rainfall_intensity_if_used = 0.0)
    
    #what is the discharge at the outlet?    
    q_at_outlet = mg.at_node['surface_water_inflow__discharge'][outlet_id]

    hydrograph_time.append(elapsed_time/3600.)
    discharge_at_outlet.append(q_at_outlet)
                          
    elapsed_time += dt
```

**Comment by kbarnhart** [comment]
@amanaster2 

I've been able to reproduce your error with the following minimally complete example:

```python
from landlab.components import KinwaveImplicitOverlandFlow, FastscapeEroder
from landlab import RasterModelGrid

#create 100x100 grid and add a slanted surface elevation
mg = RasterModelGrid(5, 5, spacing = (1,1))
z = mg.add_field('topographic__elevation', mg.node_y.copy(), at = 'node')

#initialize components
knwv = KinwaveImplicitOverlandFlow(mg)
sp = FastscapeEroder(mg, K_sp = 1.0)

# try running the kinimatic wave and fastscape components
knwv.run_one_step(1.0)
sp.run_one_step(dt = 1.0)
```

I have a few ideas about why this is occurring and anticipate being able to figure out today or tomorrow. Will keep you posted.

**Comment by kbarnhart** [comment]
@amanaster2, I think I've found the problem and fixed it in the PR listed above (PR #510).

Would you mind checking out that branch and testing the updated code on your computer with both the minimal example and the script that uses the DEM on your computer.

Let me know if it works.

**Comment by amanaster2** [comment]
@kbarnhart I will check this out and let you know! Thanks for finding and working on the problem!

**Comment by kbarnhart** [comment]
@amanaster2 did this work for you?

**Comment by amanaster2** [comment]
@kbarnhart Sorry for the delayed reply -- yes, this fix worked!

**Comment by kbarnhart** [comment]
@amanaster2 great! glad to hear!

**Comment by kbarnhart** [comment]
@amanaster2 this should have been fixed with PR #510 which has now been merged into master. 

Would you verify that this is no longer a problem for you in master and then close the issue?

**Comment by amanaster2** [comment]
@kbarnhart Sorry, this got lost among many emails in my inbox... This fix worked -- no longer a problem!
